File: services/weight_trend.py
"""
Advanced weight trend analysis.
Port of Juan-Tracker's weight_trend_calculator.dart.
Implements EMA, Holt-Winters, phase detection, and predictions.
"""

from dataclasses import dataclass
from enum import Enum


class WeightPhase(Enum):
    LOSING = "losing"
    MAINTAINING = "maintaining"
    GAINING = "gaining"
    INSUFFICIENT_DATA = "insufficient_data"


@dataclass
class TrendConfig:
    ema_period: int = 7
    hw_period: int = 7
    phase_change_threshold: float = 0.2  # kg/week
    plateau_min_days: int = 14


@dataclass
class WeightTrendResult:
    # EMA
    ema_weight: float
    ema_history: list

    # Holt-Winters
    hw_level: float
    hw_trend: float  # kg/day
    hw_prediction_7d: float = None
    hw_prediction_30d: float = None

    # Phase
    phase: WeightPhase = WeightPhase.INSUFFICIENT_DATA
    days_in_phase: int = 0
    weekly_rate: float = 0.0  # kg/week

    # Data
    latest_weight: float = 0.0

    @property
    def trend_weight(self):
        return self.ema_weight


class WeightTrendCalculator:
    def __init__(self, config=None):
        self.config = config or TrendConfig()

    def calculate(self, weight_entries):
        """
        Calculate trend from weight entries.
        weight_entries: list of dicts with 'fecha' and 'peso' keys, sorted by date.
        Returns WeightTrendResult.
        """
        weights = [e["peso"] for e in weight_entries if e.get("peso") is not None]

        if not weights:
            return WeightTrendResult(
                ema_weight=0, ema_history=[], hw_level=0, hw_trend=0,
                latest_weight=0, phase=WeightPhase.INSUFFICIENT_DATA,
            )

        if len(weights) == 1:
            return WeightTrendResult(
                ema_weight=weights[0], ema_history=weights[:],
                hw_level=weights[0], hw_trend=0,
                latest_weight=weights[0], phase=WeightPhase.INSUFFICIENT_DATA,
            )

        # EMA
        ema_history = self._calculate_ema(weights)
        ema_weight = ema_history[-1]

        # Holt-Winters
        hw_level, hw_trend = self._calculate_holt_winters(weights)

        # Predictions
        hw_prediction_7d = hw_level + hw_trend * 7
        hw_prediction_30d = hw_level + hw_trend * 30

        # Phase detection
        weekly_rate = hw_trend * 7
        phase, days_in_phase = self._detect_phase(weights, weekly_rate)

        return WeightTrendResult(
            ema_weight=ema_weight,
            ema_history=ema_history,
            hw_level=hw_level,
            hw_trend=hw_trend,
            hw_prediction_7d=hw_prediction_7d,
            hw_prediction_30d=hw_prediction_30d,
            phase=phase,
            days_in_phase=days_in_phase,
            weekly_rate=weekly_rate,
            latest_weight=weights[-1],
        )

    def _calculate_ema(self, weights):
        """Exponential Moving Average."""
        alpha = 2.0 / (self.config.ema_period + 1)
        ema = [weights[0]]
        for w in weights[1:]:
            ema.append(alpha * w + (1 - alpha) * ema[-1])
        return ema

    def _calculate_holt_winters(self, weights):
        """
        Double exponential smoothing (Holt-Winters without seasonality).
        Returns (level, trend_per_day).
        """
        if len(weights) < 2:
            return weights[0] if weights else 0, 0

        alpha = 2.0 / (self.config.hw_period + 1)
        beta = 0.3  # Trend smoothing

        level = weights[0]
        trend = weights[1] - weights[0]

        for w in weights[1:]:
            prev_level = level
            level = alpha * w + (1 - alpha) * (level + trend)
            trend = beta * (level - prev_level) + (1 - beta) * trend

        return level, trend

    def _detect_phase(self, weights, weekly_rate):
        """Detect current weight phase."""
        threshold = self.config.phase_change_threshold

        if len(weights) < 7:
            return WeightPhase.INSUFFICIENT_DATA, 0

        if weekly_rate < -threshold:
            phase = WeightPhase.LOSING
        elif weekly_rate > threshold:
            phase = WeightPhase.GAINING
        else:
            phase = WeightPhase.MAINTAINING

        # Count days in phase (simplified: use last N weights)
        days = 0
        for i in range(len(weights) - 1, max(0, len(weights) - 90), -1):
            if i == 0:
                break
            local_rate = (weights[i] - weights[i - 1]) * 7
            if phase == WeightPhase.LOSING and local_rate >= threshold:
                break
            elif phase == WeightPhase.GAINING and local_rate <= -threshold:
                break
            elif phase == WeightPhase.MAINTAINING and abs(local_rate) > threshold:
                break
            days += 1

        return phase, max(days, 1)

File: next-app/src/lib/algorithms/weight-trend.ts
/**
 * Advanced weight trend analysis.
 * Port of services/weight_trend.py (originally from Juan-Tracker).
 * Implements EMA, Holt-Winters, phase detection, and predictions.
 */

export type WeightPhase = "losing" | "maintaining" | "gaining" | "insufficient_data";

export interface TrendConfig {
  emaPeriod: number;
  hwPeriod: number;
  phaseChangeThreshold: number; // kg/week
}

const DEFAULT_CONFIG: TrendConfig = {
  emaPeriod: 7,
  hwPeriod: 7,
  phaseChangeThreshold: 0.2,
};

export interface WeightTrendResult {
  emaWeight: number;
  emaHistory: number[];
  hwLevel: number;
  hwTrend: number; // kg/day
  hwPrediction7d: number;
  hwPrediction30d: number;
  phase: WeightPhase;
  daysInPhase: number;
  weeklyRate: number; // kg/week
  latestWeight: number;
  trendWeight: number;
}

export function calculateWeightTrend(
  weightEntries: { fecha: string; peso: number | null }[],
  config: TrendConfig = DEFAULT_CONFIG
): WeightTrendResult {
  const weights = weightEntries
    .filter((e) => e.peso != null)
    .map((e) => e.peso as number);

  if (weights.length === 0) {
    return {
      emaWeight: 0,
      emaHistory: [],
      hwLevel: 0,
      hwTrend: 0,
      hwPrediction7d: 0,
      hwPrediction30d: 0,
      phase: "insufficient_data",
      daysInPhase: 0,
      weeklyRate: 0,
      latestWeight: 0,
      trendWeight: 0,
    };
  }

  if (weights.length === 1) {
    return {
      emaWeight: weights[0],
      emaHistory: [weights[0]],
      hwLevel: weights[0],
      hwTrend: 0,
      hwPrediction7d: weights[0],
      hwPrediction30d: weights[0],
      phase: "insufficient_data",
      daysInPhase: 0,
      weeklyRate: 0,
      latestWeight: weights[0],
      trendWeight: weights[0],
    };
  }

  // EMA
  const emaHistory = calculateEMA(weights, config.emaPeriod);
  const emaWeight = emaHistory[emaHistory.length - 1];

  // Holt-Winters
  const [hwLevel, hwTrend] = calculateHoltWinters(weights, config.hwPeriod);

  // Predictions
  const hwPrediction7d = hwLevel + hwTrend * 7;
  const hwPrediction30d = hwLevel + hwTrend * 30;

  // Phase detection
  const weeklyRate = hwTrend * 7;
  const { phase, daysInPhase } = detectPhase(weights, weeklyRate, config.phaseChangeThreshold);

  return {
    emaWeight,
    emaHistory,
    hwLevel,
    hwTrend,
    hwPrediction7d,
    hwPrediction30d,
    phase,
    daysInPhase,
    weeklyRate,
    latestWeight: weights[weights.length - 1],
    trendWeight: emaWeight,
  };
}

function calculateEMA(weights: number[], period: number): number[] {
  const alpha = 2.0 / (period + 1);
  const ema = [weights[0]];
  for (let i = 1; i < weights.length; i++) {
    ema.push(alpha * weights[i] + (1 - alpha) * ema[ema.length - 1]);
  }
  return ema;
}

function calculateHoltWinters(
  weights: number[],
  period: number
): [number, number] {
  if (weights.length < 2) {
    return [weights.length > 0 ? weights[0] : 0, 0];
  }

  const alpha = 2.0 / (period + 1);
  const beta = 0.3;

  let level = weights[0];
  let trend = weights[1] - weights[0];

  for (let i = 1; i < weights.length; i++) {
    const prevLevel = level;
    level = alpha * weights[i] + (1 - alpha) * (level + trend);
    trend = beta * (level - prevLevel) + (1 - beta) * trend;
  }

  return [level, trend];
}

function detectPhase(
  weights: number[],
  weeklyRate: number,
  threshold: number
): { phase: WeightPhase; daysInPhase: number } {
  if (weights.length < 7) {
    return { phase: "insufficient_data", daysInPhase: 0 };
  }

  let phase: WeightPhase;
  if (weeklyRate < -threshold) {
    phase = "losing";
  } else if (weeklyRate > threshold) {
    phase = "gaining";
  } else {
    phase = "maintaining";
  }

  // Count days in phase
  let days = 0;
  for (let i = weights.length - 1; i > Math.max(0, weights.length - 90); i--) {
    if (i === 0) break;
    const localRate = (weights[i] - weights[i - 1]) * 7;
    if (phase === "losing" && localRate >= threshold) break;
    if (phase === "gaining" && localRate <= -threshold) break;
    if (phase === "maintaining" && Math.abs(localRate) > threshold) break;
    days++;
  }

  return { phase, daysInPhase: Math.max(days, 1) };
}
